modules/frontend/deduper.go

```diff
@@ -0,0 +1,176 @@
+package frontend
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/golang/protobuf/proto"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/grafana/tempo/pkg/tempopb"
+	v1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+)
+
+const (
+	warningTooManySpans = "cannot assign unique span ID, too many spans in the trace"
+)
+
+var (
+	maxSpanID uint64 = 0xffffffffffffffff
+)
+
+func Deduper(logger log.Logger) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return spanIDDeduper{
+			next:   next,
+			logger: logger,
+		}
+	})
+}
+
+// This is copied over from Jaeger and modified to work for OpenTelemetry Trace data structure
+// https://github.com/jaegertracing/jaeger/blob/12bba8c9b91cf4a29d314934bc08f4a80e43c042/model/adjuster/span_id_deduper.go
+type spanIDDeduper struct {
+	next      Handler
+	logger    log.Logger
+	trace     *tempopb.Trace
+	spansByID map[uint64][]*v1.Span
+	maxUsedID uint64
+}
+
+// Do implements Handler
+func (s spanIDDeduper) Do(req *http.Request) (*http.Response, error) {
+	ctx := req.Context()
+	span, _ := opentracing.StartSpanFromContext(ctx, "frontend.DedupeSpanIDs")
+	defer span.Finish()
+
+	resp, err := s.next.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		body, err := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
+		if err != nil {
+			return nil, err
+		}
+
+		traceObject := &tempopb.Trace{}
+		err = proto.Unmarshal(body, traceObject)
```

Comment:
dang, sad about this. this will cost us on the big traces, but i guess there's no other way to do this using the http pipeline stuff? I wonder if Cortex/Loki repeatedly marshal/unmarshal?

is there a way to sanely combine this and the sharding pipeline processor so we only have to unmarshal/marshal once?

we want to keep the logic isolated, but it's ok if we technically use the same pipeline processor if it saves us on latency.

Reply:
A round of unmarshalling => marshalling is needed anyway in either case at the frontend regardless of where the processor sits. If we were to combine it with the querysharder, the same logic would go here - https://github.com/grafana/tempo/pull/687/commits/4bd976a4563f618a40d81a80d2ffe635ade21f4a#diff-b79ae30491380fa96485ab56b95197e2ea7f39b0f1460b7ff9d4fdeadb370ed1R198

The current approach of using the deduper as a middleware should not have a significant performance impact. 

A potentially important thing to discuss would be whether we want to make the whole module configurable so it can be turned on/off per-tenant.